File: 572-subtree-of-another-tree/572-subtree-of-another-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool IsSubtree(TreeNode root, TreeNode subRoot) {
        
        if(root == null)
            return false;
        
        if(isSameTree(root,subRoot))
            return true;
        
        return IsSubtree(root.left,subRoot) || IsSubtree(root.right,subRoot);
        
    }
    
    private bool isSameTree(TreeNode root1 , TreeNode root2)
    {
        
        if(root1 == null || root2 == null)
            return root1 == root2;
                        
        return (root1.val == root2.val) && isSameTree(root1.left,root2.left) && isSameTree(root1.right,root2.right);
                
    }
    
}

File: 572.subtree-of-another-tree.java
/*
 * @lc app=leetcode id=572 lang=java
 *
 * [572] Subtree of Another Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        
          if(root == null)
            return false;
        
          if(isSame(root,subRoot))
            return true;

        return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
        
    }

    private boolean isSame(TreeNode p, TreeNode q)
    {

        if(p == null || q == null)
           return p == null && q == null;


        return p.val == q.val && isSame(p.left,q.left) && isSame(p.right,q.right);
    }
}
// @lc code=end

